### PR TITLE
start uop docs

### PR DESCRIPTION
**Pull request by Qazalin**
http://localhost:8000/developer/uop/
image

### CODE REVIEW

File: tinygrad/ops.py [hunk 1]
```diff
@@ -40,20 +40,125 @@ def identity_element(op:BinaryOps, dt:DType): return dtypes.as_const({BinaryOps.
...
   # memory/assignment ops
-  LOAD = auto(); STORE = auto(); PHI = auto() # noqa: E702
+  LOAD = auto()
+  """
+  - **`dtype`**: Output DType
+  - **`src`**:
+      - Normal LOAD: `Tuple[UOp, UOp]`
+          - Buffer UOp `DEFINE_GLOBAL`
+          - Indexing Op, can only return dtypes.int32.
+      - Gated LOAD: `Tuple[UOp, UOp, UOp]`
```
Comment (geohot) [inline tinygrad/ops.py:121]: Gated is 4

File: tinygrad/ops.py [hunk 2]
```diff
@@ -40,20 +40,125 @@ def identity_element(op:BinaryOps, dt:DType): return dtypes.as_const({BinaryOps.
...
-  DEFINE_GLOBAL = auto(); DEFINE_VAR = auto(); DEFINE_LOCAL = auto(); DEFINE_ACC = auto() # noqa: E702
-  CONST = auto(); SPECIAL = auto() # noqa: E702
-  NOOP = auto(); GEP = auto() # noqa: E702
+  # uops that aren't rendered
+  SINK = auto()
+  """
+  Holds [UOps.STORE](/developer/uop#tinygrad.ops.UOps.STORE). SINK defines the AST for a Kernel.
+
+  - **`dtype`**: `None`
+  - **`src`**: `[UOps.STORE, ...]`
```
Comment (geohot) [inline tinygrad/ops.py:49]: src is UOps.DEFINE_GLOBAL/DEFINE_LOCAL

**Comment by geohot** [comment]
Ideally there's some way to make this a spec that actually type checks at runtime.

File: tinygrad/ops.py (hunk 2 above)
Comment (Qazalin) [reply tinygrad/ops.py:49]: sink?

**Comment by Qazalin** [comment]
I think it's possible to write a custom autogen: type_verify assert -> this structure:
```
  - **`dtype`**:  
  - **`src`**: 
  - **`arg`**: 
  ```
Wanted to start with this to move forward with a correct UOps.SWIZZLE.

File: tinygrad/ops.py (hunk 2 above)
Comment (geohot) [reply tinygrad/ops.py]: oh my bad

**Comment by geohot** [comment]
Oh, I thought they were before the define

**Comment by geohot** [comment]
Screenshot 2024-08-26 at 10 14 30 PM

Seeing this

File: tinygrad/ops.py [hunk 3]
```diff
@@ -40,20 +40,129 @@ def identity_element(op:BinaryOps, dt:DType): return dtypes.as_const({BinaryOps.
...
+  """
+  - **`dtype`**: `None`
+  - **`src`**:
+    - Normal STORE: `Tuple[UOp, UOp, UOp]`
+      - Buffer UOp [`UOps.DEFINE_GLOBAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_GLOBAL)
+        or [`UOps.DEFINE_LOCAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_LOCAL).
+      - Indexing Op, can only return `dtypes.int32`.
+      - Value to store.
+    - Gated STORE: `Tuple[UOp, UOp, UOp, UOp]`
+      - Buffer UOp [`UOps.DEFINE_GLOBAL`](/developer/uop#tinygrad.ops.UOps.DEFINE_GLOBAL).
```
Comment (geohot) [inline tinygrad/ops.py:145]: Gated local allowed?

File: tinygrad/ops.py [hunk 4]
```diff
@@ -40,20 +40,129 @@ def identity_element(op:BinaryOps, dt:DType): return dtypes.as_const({BinaryOps.
...
+  BITCAST = auto()
+  """
+  - **`dtype`**: The bitcasted scalar DType
+  - **`src`**: `Tuple[UOp]`
+  - **`arg`**: `None`
+  """
+  VECTORIZE = auto()
+  """
+  - **`dtype`**: The upcasted vector DType
+  - **`src`**: `Tuple[UOp]`
```
Comment (geohot) [inline tinygrad/ops.py:98]: Tuple[UOp, ...]?

**Comment by geohot** [comment]
I'm fine with merging this to start, but we need a way to dynamically parse and typecheck with this

**Comment by Qazalin** [comment]
hmm, getting wrong links when applying those warnings' suggestions:
image.
maybe you can only link members with other members via absolute links...

File: tinygrad/ops.py (hunk 3 above)
Comment (Qazalin) [reply tinygrad/ops.py]: my bad, can be local in multi reduce + pad.

**Comment by Qazalin** [comment]
LOAD is the longest one http://localhost:8000/developer/uop/#tinygrad.ops.UOps.LOAD
should be pretty comprehensive now.
image

**Comment by geohot** [comment]
Good to merge when you think it's ready

**Pull request merged**